### PR TITLE
support follow mode in sim

### PR DESCRIPTION
**Pull request by ariovistus**
write usable values to hal sim

### CODE REVIEW

File: gen/gen_sim.py.j2 [hunk 1]
```diff
@@ -42,3 +42,14 @@ class {{ module_name }}{% if inherits is defined %}({{ inherits }}){% endif %}:
 
   {% endif %}
   {% endfor %}
+
+  {% if module_name == 'MotController' %}
```
Comment (virtuald) [inline gen/gen_sim.py.j2]: This isn't the right way to do this. Add a 'sim_class_extra' key to MotController_data.yml and put this in the yml file. (see `sim_extra` above)

File: gen/gen_sim.py.j2 [hunk 2]
```diff
@@ -42,3 +42,14 @@ class {{ module_name }}{% if inherits is defined %}({{ inherits }}){% endif %}:
...
+    def _is_follower(self, id):
+        from ctre._impl.autogen.ctre_sim_enums import ControlMode
```
Comment (virtuald) [inline gen/gen_sim.py.j2]: Add this import to `sim_extra`? And it should be `from .ctre_sim_enums import ControlMode`.

File: gen/MotController_data.yml [hunk 3]
```diff
@@ -61,8 +62,19 @@ c_MotController_GetDescription:
...
     self.hal_data['control_mode'] = mode
-    self.hal_data['value'] = demand0
+    if mode == ControlMode.PercentOutput:
+        self.hal_data['value'] = demand0 / 1023.
+    elif mode == ControlMode.Follower:
+        self.hal_data['follow_target'] = demand0 & 0xFF
+    else:
+        self.hal_data['value'] = demand0
+
+    followers = [id for id in hal_data['CAN'].keys() if self._is_follower(id)]
```
Comment (virtuald) [inline gen/MotController_data.yml]: This computation is pretty intense. Why not save some effort and just store a list of followers?

File: gen/MotController_data.yml [hunk 4]
```diff
@@ -61,8 +62,19 @@ c_MotController_GetDescription:
     raise NotImplementedError
 c_MotController_SetDemand:
   code: |
+    from ctre._impl.autogen.ctre_sim_enums import ControlMode
     self.hal_data['control_mode'] = mode
-    self.hal_data['value'] = demand0
+    if mode == ControlMode.PercentOutput:
+        self.hal_data['value'] = demand0 / 1023.
+    elif mode == ControlMode.Follower:
+        self.hal_data['follow_target'] = demand0 & 0xFF
```
Comment (virtuald) [inline gen/MotController_data.yml]: Shouldn't you set the current value to whatever the master's current value is?

File: gen/MotController_data.yml [hunk 5]
```diff
@@ -46,6 +48,8 @@ c_MotController_Create1:
       
       'neutral_mode': self.NeutralMode.EEPROMSetting,
       
+      'follow_target': 0,
+      'followers': [],
```
Comment (virtuald) [inline gen/MotController_data.yml]: Hm, should this be a set?

File: gen/MotController_data.yml [hunk 6]
```diff
@@ -61,8 +65,24 @@ c_MotController_GetDescription:
...
+        assert follow_target != self._deviceId
+        self.hal_data['follow_target'] = follow_target
+        hal_data['CAN'][follow_target]['followers'].append(self._deviceId)
+        self.hal_data['value'] = hal_data['CAN'][follow_target]['value']
+    else:
+        self.hal_data['value'] = demand0
+
+    for follower_id in self.hal_data['followers']:
+        hal_data['CAN'][follower_id]['value'] = self.hal_data['value']
+    return retval
```
Comment (virtuald) [inline gen/MotController_data.yml]: Don't return retval, it automatically gets returned for you (and ideally will get type checked too). I should probably add something to hooks.py to check for this...

**Comment by ruwix** [comment]
This still appears to be broken as I get a KeyError setting ControlMode.Follower and running unit tests

**Pull request merged**